### DISCUSSION TITLE
Is there a way to run this with Docker?

### DISCUSSION THREAD

**Discussion by rubslopes** [3mora2/WPP_Whatsapp#37] [Q&A] [answered]
When I try to run with Docker, Playwright complains that there's no XServer running and I should set Playwright to `headless: true`

Is there an official or straightforward way to run this headless inside Docker?

Thank you,

**Comment by rubslopes** [comment, marked as answer]
I found the answer:

Add Playwright and its dependencies to the docker file:

```docker
RUN pip install playwright==1.42.0 && \
    playwright install --with-deps

RUN apt-get update && apt-get install -y \
    gconf-service \
    libasound2 \
    libatk1.0-0 \
    libcairo2 \
    libcups2 \
    libfontconfig1 \
    libgdk-pixbuf2.0-0 \
    libgtk-3-0 \
    libnspr4 \
    libpango-1.0-0 \
    libxss1 \
    fonts-liberation \
    libappindicator1 \
    libnss3 \
    lsb-release \
    xdg-utils
```

For some reason, running `playwright install --with-deps` only, without the rest, gives me a buggy performance. Not all messages are sent. But with the `apt-get` for the dependencies, it runs fine.

and, in the creator object inside your `main.py` file, use firefox and define it as headless:
```python
creator = Create(session=your_session_name, browser='firefox', headless=True)
```

For some reason I couldn't make it headless with Chrome, and someone else had already complained about that: https://github.com/3mora2/WPP_Whatsapp/issues/15#issuecomment-1817646109

I didn't know to insert the QR Code in headless mode, so I ran once without Docker with `headless=False`, read the QR Code, let it create the cookie (`tokens` folder), and then ran it headless through Docker.

**Reply by 3mora2** [reply]
You can can redirect qrcode to function and print it or show as image

**Comment by rubslopes** [comment]
> You can can redirect qrcode to function and print it or show as image

Sounds awesome! Is there a native argument of some function that prints the qrcode data? How would I got about to do that? Thank you!

**Comment by rubslopes** [comment]
Nevermind, found it!

The solution is in `examples/get_qrcode.py`, if anyone comes looking for it.

**Reply by 3mora2** [reply]
[Link](https://github.com/3mora2/WPP_Whatsapp/blob/main/examples/get_qrcode.py)
